client.go

```diff
@@ -164,6 +168,22 @@ func (c *Client) OnConnected(f OnConnectedFunc) {
 	c.onConnectedFuncs = append(c.onConnectedFuncs, f)
 }
 
+// OnError can be used to hook into the errors that the client encounters
+// outside the normal request/response cycle. Note that it is not possible to
+// call [Client.Stop] inside f. If you want to stop the client from here, you
+// should use a goroutine. This is because [Client.Stop] will block until the
+// client has stopped.
+func (c *Client) OnError(f OnErrorFunc) {
+	c.onErrorFuncs = append(c.onErrorFuncs, f)
+}
+
+// notifyOnErrorFuncs will notify everyone who listens to the [Client.OnError]
+func (c *Client) notifyOnErrorFuncs(err error) {
+	for _, onError := range c.onErrorFuncs {
+		onError(err)
+	}
+}
```

Comment:
nit: The `On` and `With` functions feels related to the construction which this does not, maybe we can move it to other regular client associate functions?

Reply:
Agreed!